architecture/core.rst

```diff
@@ -800,42 +800,52 @@ Shared Memory
 Shared Memory is a block of memory that is shared between the non-secure and the
 secure world. It is used to transfer data between both worlds.
 
-Shared Memory Allocation
+Shared Memory References
 ========================
-The shared memory is allocated by the Linux driver from a pool ``struct
-shm_pool``, the pool contains:
 
-    - The physical address of the start of the pool
-    - The size of the pool
-    - Whether or not the memory is cached
-    - List of chunk of memory allocated.
+The shared memory is allocated and managed by the non-secure world, i.e the
+Linux OP-TEE driver. Secure world only considers the individual shared buffers,
+not their pool. Each shared memory is referenced with associated attributes:
+    - Buffer start address and byte size
+    - Cacheability of the buffer
+    - List of chunks if mapped from discontiguos pages.
 
-.. note::
-    - The shared memory pool is physically contiguous.
-    - The shared memory area is **not secure** as it is used by both non-secure
-      and secure world.
+Shared memory buffer references manipulated must fit inside one of the
+shared memory areas known from the OP-TEE core. OP-TEE supports 2 kind
+of shared memory areas: a mandatory one for physically contiguous buffers
+and  optional extra memory areas for non-contiguous buffers.
+
+The shared memory area accessed with non-secure rights by OP-TEE core
+and non-secure world.
+
+Contiguous shared buffers
+-------------------------
 
-Shared Memory Configuration
-===========================
-It is the Linux kernel driver for OP-TEE that is responsible for initializing
-the shared memory pool, given information provided by the OP-TEE core. The
-Linux driver issues a SMC call ``OPTEE_SMC_GET_SHM_CONFIG`` to retrieve the
-information
+Configuration directives ``CFG_SHMEM_START`` and ``CFG_SHMEM_SIZE``
+define a share memory area where shared memory buffers are contiguous.
+Generic memory layout registers it as the ``MEM_AREA_NSEC_SHM`` memory area.
+
+The non-secure world issues ``OPTEE_SMC_GET_SHM_CONFIG`` to retrieve
+physically contiguous shared memory configuration:
 
     - Physical address of the start of the pool
     - Size of the pool
     - Whether or not the memory is cached
 
-The shared memory pool configuration is platform specific. The memory mapping,
-including the area ``MEM_AREA_NSEC_SHM`` (shared memory with non-secure world),
-is retrieved by calling the platform-specific function ``bootcfg_get_memory()``.
-Please refer to this function and the area type ``MEM_AREA_NSEC_SHM`` to see the
-configuration for the platform of interest. The Linux driver will then
-initialize the shared memory pool accordingly.
+Non-contiguous shared buffers
+-----------------------------
 
-.. todo::
+To benefit from non-contigous shared memory buffers, secure world register
+dynamic shared memory areas and non-secure world must register non-contiguous
+buffers prior referring to in the OP-TEE API.
```

Comment:
s/prior referring to in the OP-TEE API./prior to referring to them using the OP-TEE API/

Reply:
thanks for the proposal.